src/sources/pulsar.rs

```diff
@@ -0,0 +1,451 @@
+use vector_common::internal_event::{
+    BytesReceived, InternalEventHandle as _, Protocol,
+};
+use vector_common::sensitive_string::SensitiveString;
+use vector_common::shutdown::ShutdownSignal;
+use vector_config::component::GenerateConfig;
+use vector_config_macros::configurable_component;
+use vector_core::config::{LogNamespace, Output, SourceAcknowledgementsConfig};
+use vector_core::event::Event;
+use codecs::StreamDecodingError;
+use vector_core::ByteSizeOf;
+
+/// Configuration for the `pulsar` source.
+#[configurable_component(source("pulsar"))]
+#[derive(Clone, Debug, Derivative)]
+#[derivative(Default)]
+#[serde(deny_unknown_fields)]
+pub struct PulsarSourceConfig {
+    /// The endpoint to which the Pulsar client should connect to.
+    #[serde(alias = "address")]
+    endpoint: String,
+
+    /// The Pulsar topic names to read events from.
+    topics: Vec<String>,
+
+    /// The Pulsar consumer name.
+    consumer_name: Option<String>,
+
+    /// The Pulsar subscription name.
+    subscription_name: Option<String>,
+
+    /// Priority level for a consumer to which a broker gives more priority while dispatching messages in Shared subscription type.
+    ///
+    /// The broker follows descending priorities. For example, 0=max-priority, 1, 2,...
+    ///
+    /// In Shared subscription type, the broker first dispatches messages to the max priority level consumers if they have permits. Otherwise, the broker considers next priority level consumers.
+    priority_level: Option<i32>,
+
+    /// Max count of messages in a batch.
+    batch_size: Option<u32>,
+
+    #[configurable(derived)]
+    auth: Option<AuthConfig>,
+
+    #[configurable(derived)]
+    dead_letter_queue_policy: Option<DeadLetterQueuePolicy>,
+
+    #[configurable(derived)]
+    #[serde(default = "default_framing_message_based")]
+    #[derivative(Default(value = "default_framing_message_based()"))]
+    framing: FramingConfig,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    #[derivative(Default(value = "default_decoding()"))]
+    decoding: DeserializerConfig,
+
+    #[configurable(derived)]
+    #[serde(default, deserialize_with = "bool_or_struct")]
+    acknowledgements: SourceAcknowledgementsConfig,
+}
+
+/// Authentication configuration.
+#[configurable_component]
+#[derive(Clone, Debug)]
+struct AuthConfig {
+    /// Basic authentication name/username.
+    ///
+    /// This can be used either for basic authentication (username/password) or JWT authentication.
+    /// When used for JWT, the value should be `token`.
+    name: Option<String>,
+
+    /// Basic authentication password/token.
+    ///
+    /// This can be used either for basic authentication (username/password) or JWT authentication.
+    /// When used for JWT, the value should be the signed JWT, in the compact representation.
+    token: Option<SensitiveString>,
+
+    #[configurable(derived)]
+    oauth2: Option<OAuth2Config>,
+}
```

Comment:
As this is shared with the source, it would be good to move bits like this into a common shared module.